lib/charms/grafana_agent/v0/cos_machine.py

```diff
@@ -0,0 +1,229 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Library for the cos_machine relation interface."""
+
+import base64
+import json
+import logging
+import lzma
+from pathlib import Path
+from typing import Dict, List, Optional, Union
+
+# FIXME: unify the alert rules format in cosl to drop these ASAP
+from charms.loki_k8s.v0.loki_push_api import AlertRules as LogAlerts
+from charms.prometheus_k8s.v0.prometheus_scrape import AlertRules as MetricsAlerts
+from ops.charm import RelationEvent
+from ops.framework import EventBase, EventSource, Object, ObjectEvents
+
+LIBID = ""
+LIBAPI = 0
+LIBPATCH = 1
+
+PYDEPS = ["cosl"]
+
+DEFAULT_RELATION_NAME = "cos-machine"
+DEFAULT_METRICS_ENDPOINT = {
+    "path": "/metrics",
+    "port": 80,
+}
+
+logger = logging.getLogger(__name__)
+
+
+class CosMachineProvider(Object):
+    """Provider class for the cos_machine interface."""
+
+    def __init__(
+        self,
+        charm,
+        relation_name: str = DEFAULT_RELATION_NAME,
+        metrics_endpoints: List[dict] = [DEFAULT_METRICS_ENDPOINT],
```

Comment:
I'm not sure about `metrics_endpoints`.... I mean, a charm that instruments our lib will need to do something like:

```python
        self._grafana_agent = CosMachineProvider(
            self, metrics_endpoints=[{"path": "/metrics", "port": "9100"}]
        )
```

But if you relate the same charm with Prometheus:

```python
        self.metrics_endpoint = MetricsEndpointProvider(
            self, jobs=[{"static_configs": [{"targets": ["*:9100", "*:9101"]}]}]
        )
```

Shouldn't we use the same pattern here?

Reply:
I'm not sure, on the one hand this way is more simple and requires the more minimal interface, but also consistency is nice what do people think?

Reply:
I'd rather go for consistency, verbosity can be scripted away, confusion can't

Reply:
I would also go for consistency... seems the more "natural" way to proceed

Reply:
Using the term `static_configs` is technically wrong for both cases because star-notation `"*:9100"`  nor `path, port` are recognized by prometheus.

I prefer `metrics_endpoints=[{"path": "/metrics", "port": "9100"}]` here.

Reply:
Maybe even `List[Tuple[int, str]]` like `[(9100, "/metrics"), ...]`

Reply:
The downside of `List[Tuple[int, str]]]` is that `metrics_endpoints[0][0]` is horrifying.

We'd be better off with `NamedTuple` or a trivial dataclass